### PR TITLE
sinkmanager(ticdc): release mem quota when remove a table

### PR DESCRIPTION
**Pull request by hi-rustin**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tiflow/issues/5928

### What is changed and how it works?

Release mem quota when removing a table.

Abort the task and refund the unused memory quota. Also, do not put the progress back into the heap.

#### Tests 

 - [x] Unit test

##### Will it cause performance regression or break compatibility?
No
##### Do you need to update user documentation, design documentation or monitoring documentation?
No
### Release note 

```release-note
None
```

### CODE REVIEW

File: cdc/processor/sinkmanager/manager_impl.go [hunk 1]
```diff
@@ -292,14 +295,17 @@ func (m *ManagerImpl) RemoveTable(tableID model.TableID) error {
 			zap.String("changefeed", m.changefeedID.ID),
 			zap.Int64("tableID", tableID))
 	}
+	// Awake all workers to check if the table sink is removed.
+	m.memQuota.awakeAll()
```
Comment (hicqu) [inline cdc/processor/sinkmanager/manager_impl.go:299]: Seems should close the table sink before awake all.

Comment (hi-rustin) [reply cdc/processor/sinkmanager/manager_impl.go:299]: Yes. Good catch!

File: cdc/processor/sinkmanager/mem_quota.go [hunk 2]
```diff
@@ -78,17 +78,18 @@ func (m *memQuota) forceAcquire(nBytes uint64) {
...
 
 // blockAcquire is used to block the request when the memory quota is not available.
-func (m *memQuota) blockAcquire(nBytes uint64) error {
+func (m *memQuota) blockAcquire(nBytes uint64) (errChan chan error) {
 	m.mu.Lock()
 	defer m.mu.Unlock()
 	for {
 		if m.isClosed.Load() {
-			return cerrors.ErrFlowControllerAborted.GenWithStackByArgs()
+			errChan <- cerrors.ErrFlowControllerAborted.GenWithStackByArgs()
```
[the remark this answers is not in the document]
Comment (hi-rustin) [reply cdc/processor/sinkmanager/mem_quota.go]: I reverted it.

File: cdc/processor/sinkmanager/mem_quota.go [hunk 3]
```diff
@@ -78,14 +78,18 @@ func (m *memQuota) forceAcquire(nBytes uint64) {
...
-func (m *memQuota) blockAcquire(nBytes uint64) error {
+func (m *memQuota) blockAcquire(nBytes uint64, isAborted func() bool) error {
 	m.mu.Lock()
 	defer m.mu.Unlock()
 	for {
 		if m.isClosed.Load() {
 			return cerrors.ErrFlowControllerAborted.GenWithStackByArgs()
 		}
 
+		if isAborted() {
```
Comment (hicqu) [inline cdc/processor/sinkmanager/mem_quota.go:89]: It's possible that here it's not aborted, but it becomes aborted immediately. Is it ok?

File: cdc/processor/sinkmanager/mem_quota.go [hunk 4]
```diff
@@ -156,6 +160,30 @@ func (m *memQuota) release(tableID model.TableID, resolved model.ResolvedTs) {
...
+		return
+	}
+	records := m.tableMemory[tableID]
+	for _, record := range records {
+		m.usedBytes -= record.size
+	}
+	delete(m.tableMemory, tableID)
+
+	if m.usedBytes < m.totalBytes {
+		m.blockAcquireCond.Signal()
```
Comment (hicqu) [inline cdc/processor/sinkmanager/mem_quota.go:183]: Why not awake all here? So the another `awake` isn't necessary.

Comment (hi-rustin) [reply cdc/processor/sinkmanager/mem_quota.go:183]: I think we have to stop completely before we can proceed to clean.

**Review by nongfushanquan — APPROVED** [review]

File: cdc/processor/sinkmanager/mem_quota.go [hunk 5]
```diff
@@ -156,6 +156,30 @@ func (m *memQuota) release(tableID model.TableID, resolved model.ResolvedTs) {
...
```
Comment (hicqu) [inline cdc/processor/sinkmanager/mem_quota.go:179]: How about always notifing all?

**Comment by hicqu** [comment]
Rest LGTM.

File: cdc/processor/sinkmanager/mem_quota.go (hunk 5 above)
Comment (hicqu) [reply cdc/processor/sinkmanager/mem_quota.go:179]: For example:
```go
if prevUsedBytes >= m.totalBytes && m.usedBytes < m.totalBytes {
   m.blockAcquireCond.NotifyAll()
}
```

Comment (hi-rustin) [reply cdc/processor/sinkmanager/mem_quota.go]: Updated!

File: cdc/processor/sinkmanager/mem_quota.go [hunk 6]
```diff
@@ -156,6 +156,37 @@ func (m *memQuota) release(tableID model.TableID, resolved model.ResolvedTs) {
...
+		return
+	}
+	cleaned := uint64(0)
+	records := m.tableMemory[tableID]
+	for _, record := range records {
+		cleaned += record.size
+	}
+	m.usedBytes -= cleaned
+	delete(m.tableMemory, tableID)
+	log.Debug("MemoryQuotaTracing: Clean table memory records",
```
Comment (hicqu) [inline cdc/processor/sinkmanager/mem_quota.go:179]: Seems other mem quota tracing logs are in callers and after the memory quota operation finished. How about make this functions returns `cleanedBytes`, and log the message in callers?

Comment (hi-rustin) [reply cdc/processor/sinkmanager/mem_quota.go]: Updated! Thanks! Good suggestion.

**Review by hicqu — APPROVED** [review]

**Pull request merged**